graylog2-server/src/main/java/org/graylog/plugins/views/search/rest/ViewsResource.java

```diff
@@ -139,15 +150,67 @@ public ViewDTO get(@ApiParam(name = "id") @PathParam("id") @NotEmpty String id)
     @POST
     @ApiOperation("Create a new view")
     @AuditEvent(type = ViewsAuditEventTypes.VIEW_CREATE)
-    public ViewDTO create(@ApiParam @Valid ViewDTO dto, @Context UserContext userContext) throws ValidationException {
+    public ViewDTO create(@ApiParam @Valid  @NotNull(message = "View is mandatory") ViewDTO dto, @Context UserContext userContext) throws ValidationException {
         if (dto.type().equals(ViewDTO.Type.DASHBOARD)) {
             checkPermission(RestPermissions.DASHBOARDS_CREATE);
         }
+
+        validateIntegrity(dto);
+
         final User user = userContext.getUser();
         final ViewDTO savedDto = dbService.saveWithOwner(dto.toBuilder().owner(user.getName()).build(), user);
         return savedDto;
     }
 
+    private void validateIntegrity(ViewDTO dto) {
+        final Search search = searchDbService.get(dto.searchId()).orElseThrow(() -> new BadRequestException("Search " + dto.searchId() + " doesn't exist"));
+
+        final Set<String> searchQueries = search.queries().stream()
+                .map(Query::id)
+                .collect(Collectors.toSet());
+
+        final Set<String> stateQueries = dto.state().keySet();
+
+        if(!searchQueries.equals(stateQueries)) {
+            throw new BadRequestException("Search queries do not correspond to view/state queries");
+        }
+
+        final Set<String> searchTypes = search.queries().stream()
+                .flatMap(q -> q.searchTypes().stream())
+                .map(SearchType::id)
+                .collect(Collectors.toSet());
+
+
+        final Set<String> stateTypes = dto.state().values().stream()
+                .flatMap(v -> v.widgetMapping().values().stream())
+                .flatMap(Collection::stream)
+                .collect(Collectors.toSet());
+
+        if(!searchTypes.equals(stateTypes)) {
+            throw new BadRequestException("Search types do not correspond to view/search types");
+        }
+
+        final Set<String> widgetIds = dto.state().values().stream()
+                .flatMap(v -> v.widgets().stream())
+                .map(WidgetDTO::id)
+                .collect(Collectors.toSet());
+
+        final Set<String> widgetMappings = dto.state().values().stream()
+                .flatMap(v -> v.widgetMapping().keySet().stream()).collect(Collectors.toSet());
+
+        final Set<String> widgetPositions = dto.state().values().stream()
+                .flatMap(v -> v.widgetPositions().keySet().stream()).collect(Collectors.toSet());
+
+        if(!widgetMappings.equals(widgetIds)) {
```

Comment:
This implicitly assumes that there is a 1:1 relation between widget mappings and widgets. But, we could also have:

  - a widget that does not require any search types (for example a markdown widget showing a static text or querying a remote API)
  - a widget mapping for a widget that does not exist anymore, which has not been cleaned up

While the second might indicate an underlying issue in the client's state management, both are not error cases that we should handle here.

Reply:
removed this check